packages/tenderly-hardhat/src/utils/util.ts

```diff
@@ -1,13 +1,220 @@
+  forkId: string
+): Promise<TenderlyForkContractUploadRequest | null> => {
+  const request = await _makeVerifyContractsRequest(hre, flatContracts, forkId);
+  const forkRequest: TenderlyForkContractUploadRequest = await _convertToForkRequest(
+    hre,
+    request as TenderlyVerifyContractsRequest
+  );
+
+  return {
+    ...forkRequest,
+    root: txRoot,
+  };
+};
+
+async function _makeVerifyContractsRequest(
+  hre: HardhatRuntimeEnvironment,
+  flatContracts: ContractByName[],
+  forkId?: string
+): Promise<TenderlyVerifyContractsRequest | null> {
+  logger.info("Processing data needed for verification.");
+
+  const contracts: TenderlyVerificationContract[] = [];
+  for (const flatContract of flatContracts) {
+    logger.info("Processing contract:", flatContract.name);
+
+    let job: CompilationJob;
+    try {
+      job = await getCompilationJob(hre, flatContract.name);
+    } catch (err) {
+      // TODO(dusan): See how to wrap errors, don't return errors like this
+      logger.error(
+        `Error while trying to get compilation job for contract '${flatContract.name}'. The provided contract probably doesn't exist or is mistyped.`
+      );
+      throw err;
+    }
+
+    const network = hre.hardhatArguments.network;
+    if (network === undefined) {
+      logger.error(
+        `Error in ${PLUGIN_NAME}: Please provide a network via the hardhat --network argument or directly in the contract`
+      );
+      return null;
+    }
+    logger.trace("Found network is:", network);
+
+    let chainId: string = NETWORK_NAME_CHAIN_ID_MAP[network.toLowerCase()];
+    if (hre.config.networks[network].chainId !== undefined) {
+      chainId = hre.config.networks[network].chainId!.toString();
+    }
+    if (chainId === undefined && network === "tenderly" && forkId !== undefined) {
+      chainId = forkId;
+    }
+    logger.trace(`ChainId for network '${network}' is ${chainId}`);
+
+    if (chainId === undefined) {
+      logger.error(
+        `Error in ${PLUGIN_NAME}: Couldn't identify network. Please provide a chainId in the network config object`
+      );
+      return null;
+    }
+
+    contracts.push({
+      contractToVerify: flatContract.name,
+      // TODO(dusan) this can be done with TASK_COMPILE_SOLIDITY_GET_COMPILER_INPUT hardhat task
+      sources: await extractSources(job),
+      compiler: job.getSolcConfig(),
+      networks: {
+        [chainId]: {
+          address: flatContract.address,
+          links: flatContract.libraries,
+        },
+      },
+    });
+  }
+  // TODO(dusan) see about merging compilation jobs?
+
+  return {
+    contracts,
+  };
+}
+
+async function extractSources(job: CompilationJob): Promise<Record<string, TenderlyVerifyContractsSource>> {
+  const sources: Record<string, TenderlyVerifyContractsSource> = {};
+  logger.info("Extracting sources from compilation job.");
+
+  for (const file of job.getResolvedFiles()) {
+    const name = file.sourceName.split("/").slice(-1)[0].split(".")[0];
```

Comment:
This is pretty ugly and hard to understand. Can we at least make a function for it?
`e.g. extractFileName()`

Reply:
For sure, good point. 😄